robocluster/member.py

```diff
@@ -0,0 +1,374 @@
+
+    def __init__(self, member, name, uid):
+        super().__init__(member)
+
+        self.name = name
+        self.uid = uid
+
+        self._address = None
+        self.subscriptions = set()
+
+        self._wanted = set()
+        self._is_wanted = asyncio.Event(loop=self.loop)
+
+        self._pending = {}
+
+        self._socket = None
+        self._connected = asyncio.Event(loop=self.loop)
+
+        self.add_daemon_task(self._recv_loop)
+
+    @property
+    def address(self):
+        return self._address
+
+    @address.setter
+    def address(self, new):
+        if new != self._address:
+            self.close()
+            self._address = new
+
+    @property
+    def connected(self):
+        self.member.wanted.add(self.name)
+        return self._connected.wait()
+
+    async def send(self, endpoint, data):
+        # TODO: timeout?
+        await self.connected
+        packet = 'send', (endpoint, data)
+        await self._send(packet)
+
+    async def publish(self, endpoint, data):
+        for subscription in self.subscriptions:
+            if fnmatch(endpoint, subscription):
+                await self.send(endpoint, data)
+
+    async def _handle_send(self, packet):
+        endpoint, data = packet
+        await self.member._handle_send(self.name, endpoint, data)
+
+    async def request(self, endpoint, *args, **kwargs):
+        await self.connected
+        rid = int.from_bytes(os.urandom(4), 'big')
+        packet = 'request', (rid, endpoint, args, kwargs)
+        future = self._pending[rid] = asyncio.Future(loop=self.loop)
+        await self._send(packet)
+        return await future
+
+    async def _handle_request(self, packet):
+        rid, endpoint, args, kwargs = packet
+        result = await self.member._handle_request(endpoint, *args, **kwargs)
+        packet = 'response', (rid, result)
+        await self._send(packet)
+
+    async def _handle_response(self, packet):
+        rid, result = packet
+        future = self._pending.pop(rid, None)
+        if future:
+            future.set_result(result)
+
+    async def accept(self, conn):
+        if self._socket is not None:
+            conn.close()
+        self._socket = conn
+        self._connected.set()
+
+    async def _send(self, packet):
+        packet = json.dumps(packet).encode()
+        size = len(packet).to_bytes(4, 'big')
+        await self._socket.send(size + packet)
+
+    async def _recv_loop(self):
+        member = self.member
+
+        while ...:
+            await self._is_wanted.wait()
+
+            if member.uid >= self.uid:
+                # The other side will connect to me
+                await self._connected.wait()
+            elif not self._connected.is_set():
+                # I am responsible for doing the connect!
+                self._socket = self.socket('tcp')
+
+                try:
+                    await self._socket.connect(self._address)
+                except (ConnectionRefusedError, OSError):
```

Comment:
Will we provide some way to debug this in the future, so we can tell if a connection is failing continuously?

Reply:
I should probably use the builtin logger to help us with this...